### PR TITLE
LP-2454 Write unit tests, cater missed edge cases and improve code

### PR DESCRIPTION
**Pull request by harris-irfan**
### Description

[LP-2454](https://philanthropyu.atlassian.net/browse/LP-2454)

[adg-edx-theme PR](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/104)

#### Notes for reviewers

1. A new file `openedx/adg/constants.py` is added for constants used across LMS, CMS and common apps.

2. A constant `MONTH_DAY_YEAR_FORMAT = '%m/%d/%Y'` is added in the new `constants.py` file

3. All instances with `%m/%d/%Y` are replaced with the named constant `MONTH_DAY_YEAR_FORMAT`. The existing files impacted are as follows:

  - openedx/adg/common/mailchimp_pipeline/helpers.py
  - openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py
  - openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py

4. The directory `openedx/adg/lms/applications/test` is renamed to `openedx/adg/lms/applications/tests` as per the convention and for consistency. All import paths that previously referred to `openedx.adg.lms.applications.test...` are updated to now refer to `openedx.adg.lms.applications.tests...`. The following existing files are impacted by this change:

  - openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py
  - openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py
  - openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py
  - openedx/adg/lms/applications/tests/test_handlers.py
  - openedx/adg/lms/registration_extension/tests/factories.py

5. Short descriptions are added for all field methods in `openedx/adg/lms/applications/admin.py` to make their labels localized.

6. A small functional change is made in `openedx/adg/lms/applications/admin.py.UserApplicationADGAdmin.changeform_view` to pre-populate admin note in case an admin has entered a note and submitted the form without making a decision. In such a case, the note they entered in the first place should be pre-populated when the form is re-rendered with the error message.

7. A small functional change is made in `openedx/adg/lms/applications/admin.py.UserApplicationADGAdmin._get_applicant_info_fieldset` to cater the case when an applicant has not provided their organization as part of the application. In such a case, `organization` field should not be a part of the applicant info fields and its label should not be shown.

8. Most repeated strings are replaced with named constants as there were many strings that were both used in tests and the original code. To avoid redundancy, they are moved to constants files and then imported in both the original code files and the unit test code files.

9. A small fix related to `location` field method in `openedx/adg/lms/applications/admin.py.UserApplicationADGAdmin` is made to cater the case when the country of the applicant is not stored in their user profile.

10. Changes are made in `openedx/adg/lms/applications/models.py` file and resultantly a new migration file 0004 is added.

  - The admin note field in `UserApplication` model is made non-nullable as it is a TextField and Django does not recommend using null=True for CharField or TextField
  - The order of the choices in the status field of `UserApplication` is also updated to match the designs
  - An unused `file_attached` property is removed from `UserApplication` model

11. A new file `openedx/adg/lms/applications/tests/conftest.py` is added that hosts all the common fixtures used across different test modules in `openedx/adg/lms/applications/tests`. No existing fixtures were moved to this file, rather all the new fixtures written for unit tests of LP-2454 are placed in this file.

12. This PR contains unit tests that covers each and every method/function/property that was written in the previous PR under the scope of LP-2454. 100% unit test coverage is ensured.

13. The failing "Code quality with Pylint, iSort & Pycodestyle / quality (pull_request)" check is probably because of an issue in pylint reporting. It does not show any error on my code. The quality check [fails](https://github.com/OmnipreneurshipAcademy/edx-platform/pull/140/checks?check_run_id=1801290326) to complete and returns "Error: Process completed with exit code 1."

### CODE REVIEW

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/963/workflows/workflows_id/jobs/963/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/course_meta/tests/test_models.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/constants.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/forms.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/conftest.py (100%)
openedx/adg/lms/applications/tests/constants.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_inlines.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py (100%)
openedx/adg/lms/applications/tests/test_forms.py (100%)
openedx/adg/lms/applications/tests/test_handlers.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   494 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/965/workflows/workflows_id/jobs/965/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/constants.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/forms.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/conftest.py (100%)
openedx/adg/lms/applications/tests/constants.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_inlines.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py (100%)
openedx/adg/lms/applications/tests/test_forms.py (100%)
openedx/adg/lms/applications/tests/test_handlers.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   472 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by asadali145 — APPROVED** [review]
Just a small suggestion, Great Work👏🏻

File: openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+def test_changeform_view(
+    mock_save_application_review_info,
+    mock_get_extra_context_for_application_review_page,
+    mock_changeform_view,
+    request,
+    user_application,
+    user_application_adg_admin_instance,
+    request_method,
+    status,
+    note
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py:151]: @asadali145 the idea behind this is that if the input parameters of a function fit under one line, I place them in a single line. Else, I place every parameter in a separate line. This practice, I have used consistently throughout the unit test code and the original code. This is backed by the two line break options [recommended](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#breaking-long-lines) by edx, and also internally agreed upon.

[the remark this answers is not in the document]
Comment (asadali145) [reply openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py:151]: Yes, You are right. Both are correct. I thought may be we can use 1 of these. Anyway no issues.

**Review by javeriaejaz-arbisoft — APPROVED** [review]
LGTM

File: openedx/adg/lms/applications/models.py [hunk 2]
```diff
@@ -139,21 +138,21 @@ class UserApplication(TimeStampedModel):
...
+        (ACCEPTED, _('Accepted'))
     )
     status = models.CharField(
         verbose_name=_('Application Status'), choices=STATUS_CHOICES, max_length=8, default=OPEN,
     )
     reviewed_by = models.ForeignKey(
         User, null=True, blank=True, on_delete=models.CASCADE, verbose_name=_('Reviewed By')
     )
-    internal_admin_note = models.TextField(null=True, blank=True, verbose_name=_('Admin Note'))
+    internal_admin_note = models.TextField(blank=True, verbose_name=_('Admin Note'))
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/models.py:155]: @javeriaejaz-arbisoft Django does not recommend using `null=True` for CharField or TextField. The following excerpt is taken from [Django official docs](https://docs.djangoproject.com/en/3.1/ref/models/fields/#null):

"Avoid using null on string-based fields such as CharField and TextField. If a string-based field has null=True, that means it has two possible values for “no data”: NULL, and the empty string. In most cases, it’s redundant to have two possible values for “no data;” the Django convention is to use the empty string, not NULL."

As far as `blank=True` is concerned, if I remove it, the field will be made mandatory and any QA while creating a test user application from admin will have to first add an internal note, which should not be the case. The QA needs to create user application with blank note, so they can later populate the note and run test cases related to adding notes while reviewing an application.

As far as assigning a default value is concerned, by default, the default value for a TextField or CharField is an empty string. But yes, it may cause a problem while running a migration on dev/staging/prod server if there are existing applications with internal_admin_note = null, and we might have to cater that. The problem occurred locally for me as well and I had to provide a default '' while running the migration in the command line. Well pointed out! For that, instead of remaking the migration with default='', a better approach would be to make sure that before running deployment in dev/staging/prod, there are no user application with empty internal note. We can manually populate that, since ADG isn't live yet. There are not too many applications on those servers. Only three were there in dev, and one in staging. I have updated them to avoid this issue. Explicitly providing an empty string as default may not be very favourable, because as I mentioned it is already defined for TextField in Django.

File: openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+            UserApplicationADGAdmin.changeform_view(user_application_adg_admin_instance, request, application_id)
+
+            mock_save_application_review_info.assert_not_called()
+            mock_get_extra_context_for_application_review_page.assert_called_once_with(user_application)
+            mock_changeform_view.assert_called_once_with(request, application_id, extra_context=expected_context)
+    elif request.method == 'GET':
+        UserApplicationADGAdmin.changeform_view(user_application_adg_admin_instance, request, application_id)
+
+        mock_get_extra_context_for_application_review_page.assert_called_once_with(user_application)
+        mock_changeform_view.assert_called_once_with(request, application_id, extra_context=expected_context)
```
Comment (ahmed-faraz46) [inline openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py:190]: In this particular case, for each set of parameter we have different set of assertion and we have implemented different conditions to separate out different cases. So it is not a good case to handle with parametrization. Let's please write separate test cases for each case, which would be comparatively more straightforward, easy to understand and fix, with minimal possible duplication.

File: openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+    if resume:
+        with pytest.raises(StopIteration):
+            next(actual_formsets)
+    else:
+        assert next(actual_formsets) == ('test_formset', education_inline_instance)
+        if has_work_experience:
+            assert next(actual_formsets) == ('test_formset', work_experience_inline_instance)
+        else:
+            with pytest.raises(StopIteration):
+                next(actual_formsets)
```
Comment (ahmed-faraz46) [inline openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py:610]: Same is the case with this one, where we have to implement conditions to handle different parameters, we should consider moving them to separate test case.

**Review by ahmed-faraz46 — CHANGES_REQUESTED** [review]
What a mater piece Harris 👏 Well thought and well planned with max coverage! Here are just a few minor things I noticed.

File: openedx/adg/lms/applications/admin.py [hunk 5]
```diff
@@ -230,9 +244,10 @@ def get_queryset(self, request):
 
     def applicant_name(self, obj):
         return obj.user.profile.name
+    applicant_name.short_description = _('Applicant Name')
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/admin.py:247]: If a string is used more than once, that is susceptible to change, it is kept in constants.py. All other strings that are used only once are not placed in constants. You might find some exceptions in strings used in a single test twice/thrice, that are not placed in constants. That is because a string used in a test is not susceptible to change, like 'test_formset' . However, if there is a relatively generic test string that is used in multiple test cases, like 'Test note', then it is placed in constants.py. Placing every string in constants.py would be an overkill as it will result in a very long list of imports.

File: openedx/adg/lms/applications/constants.py [hunk 6]
```diff
@@ -38,3 +39,34 @@
...
+    'f': _('Woman'),
+    'o': _('Prefer not to answer')
+}
+
+DAY_MONTH_YEAR_FORMAT = '%d %B %Y'
+MONTH_NAME_DAY_YEAR_FORMAT = '%B %d, %Y'
+
+CourseScore = namedtuple('CourseScore', 'course_name course_percentage')
+
+APPLICANT_INFO_FIELDSET_TITLE = _('APPLICANT INFORMATION')
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py:68]: You can see the comment headers to know how the constants are grouped. The other constants not falling under a specific group are added without keeping any certain order in mind.

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/975/workflows/workflows_id/jobs/975/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/constants.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/forms.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/conftest.py (100%)
openedx/adg/lms/applications/tests/constants.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_inlines.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py (100%)
openedx/adg/lms/applications/tests/test_forms.py (100%)
openedx/adg/lms/applications/tests/test_handlers.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   477 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by ahmed-faraz46 — APPROVED** [review]
Looks perfect, just re confirm, if this constant is not being used anywhere, please remove it.

File: openedx/adg/lms/applications/constants.py [hunk 7]
```diff
@@ -38,3 +41,32 @@
 RESUME_DISPLAY = 'resume_display'
 COVER_LETTER_TEXT = 'cover_letter'
 PREREQUISITES = 'prerequisites'
+
+APPLICATION_RECEPTION_DATE_FORMAT = ''
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py]: no, removed.

[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py]: Strange that it bypassed linting.

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/997/workflows/workflows_id/jobs/997/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/constants.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/forms.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (100%)
openedx/adg/lms/applications/tests/conftest.py (100%)
openedx/adg/lms/applications/tests/constants.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_inlines.py (100%)
openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py (100%)
openedx/adg/lms/applications/tests/test_forms.py (100%)
openedx/adg/lms/applications/tests/test_handlers.py (100%)
openedx/adg/lms/applications/tests/test_helpers.py (100%)
openedx/adg/lms/applications/tests/test_models.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   476 lines
Missing: 0 lines
Coverage: 100%
-------------
```
